tests/integration/main.rs

```diff
@@ -0,0 +1,93 @@
+mod artichoke;
+
+use std::env;
+use std::path::PathBuf;
+use std::process::Command;
+use serde::ser::{Serialize, Serializer, SerializeStruct};
+use bstr::BString;
+
+#[allow(dead_code)]
+#[derive(Debug)]
+struct CommandOutput {
+    call_args: Vec<String>,
+    status: i32,
+    stdout: BString,
+    stderr: BString,
+}
+
+impl CommandOutput {
+    fn new() -> Self {
+        Self {
+            call_args: vec![],
+            status: -1,
+            stdout: BString::from(""),
+            stderr: BString::from(""),
+        }
+    }
+
+    fn with_args(&mut self, call_args: &Vec<&str>) -> &mut Self {
+        self.call_args.append(&mut call_args.clone().iter().map(|x| x.to_string()).collect());
+        self
+    }
+
+    fn with_command_output(&mut self, output: std::process::Output) -> &mut Self {
+        self.status = output.status.code().unwrap_or(-1);
+        self.stdout = BString::from(output.stdout);
+        self.stderr = BString::from(output.stderr);
+        self
+    }
+
+    fn build(&self) -> Self {
+        CommandOutput {
+            call_args: self.call_args.clone(),
+            status: self.status,
+            stdout: self.stdout.clone(),
+            stderr: self.stderr.clone(),
+        }
+    }
+}
+
+impl Serialize for CommandOutput {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("parameters", 4)?;
+        s.serialize_field("call_args", &self.call_args)?;
+        s.serialize_field("status", &self.status)?;
+        s.serialize_field("stdout", &format!("{}", self.stdout))?;
+        s.serialize_field("stderr", &format!("{}", self.stderr))?;
```

Comment:
Formatting the `BString` with `Display` is not quite right since the impl replaces invalid UTF-8 with '�'.

I think we'll probably want something like this:

```suggestion
        let mut buf = String::new();
        for line in self.stdout.lines() {
            write!(&mut buf, "{:?}", line.as_bstr())?;
        }
        s.serialize_field("stdout", &buf)?;

        buf = String::new();
        for line in self.stderr.lines() {
            write!(&mut buf, "{:?}", line.as_bstr())?;
        }
        s.serialize_field("stderr", &buf)?;
```

This will turn non-printable control characters into ASCII escape sequences and invalid UTF-8 byte sequences into hex escape sequences, which keeps everything lossless but also human readable.

Reply:
I had to change a couple of things on this. At first I was confused as to why to split on lines, to convert to `Bstr`, and reassemble (e.g. just convert the BString, instead of line by line), but then I saw that `\n` was being output and combining it all into one string.

I took an alternative than a loop, since `write!` would need to be `writeln!` if we still wanted it human readable. However, in doing that, we might be adding an additional `\n` in the output, which might not be 100% correct - thus the `join` approach. Iter vs. for loop doesn't worry me, so if you prefer a for loop, let me know!

Reply:
Yup that works for me. This is similar to how another test helper for printing backtraces in functional tests works:

https://github.com/artichoke/artichoke/blob/5f70f7ee62b903bcb62ef03f0e994771da9744f3/artichoke-backend/src/test/prelude.rs#L81-L89